PR: https://github.com/odoo/odoo/pull/269357

From: 2b081090fbd3a2aebb6443e18b77cf7f693d63f3
From: Stéphane Diez
Date: 2026-09-08 16:25:02

Breaking data model changes scores: del:1 + add:0, change matches:
-    _inherit = 'stock.traceability.report'

Total Changes: 41

[REF] stock: refactor traceability report

Expand towards end products and customers:
Instead of moving downstream moving from one report to the next, we're adding a way
to find all the 'children' lines of the select stock move line(s). This way we can
stay in the same report and have a global view upstream and downstream from any
point in the chain.

Moreover, we're modifying the Unfold/Fold button to fetch all records in python
before sending the final result at once to the JS. This limits the number of calls
between JS and python to one.

If we enter the report from the middle of a chain, there will be at least two lines
show, one going upstream (parent line, existing behavior) and one going downstream
(child line, new behavior).

The most recent lines are shown at the top of the report.

task 5265405

Part-of: odoo/odoo#269357
Signed-off-by: Arnold Moyaux (arm) <arm@odoo.com>

================================= pseudo patch: =================================

--- a/addons/mrp/models/__init__.py
+++ b/addons/mrp/models/__init__.py
@@ -6,7 +6,6 @@ from . import mrp_bom
 from . import mrp_routing
 from . import mrp_workcenter
 from . import mrp_production
-from . import stock_traceability
 from . import mrp_unbuild
 from . import mrp_workorder
 from . import product

--- a/addons/mrp/models/stock_traceability.py
+++ b/None
@@ -1,40 +0,0 @@
-from odoo import models, api
-
-
-class StockTraceabilityReport(models.TransientModel):
-    _inherit = 'stock.traceability.report'
-
-    @api.model
-    def _get_reference(self, move_line):
-        res_model, res_id, ref = super()._get_reference(move_line)
-        move = move_line.move_id
-        if move.production_id and move.location_dest_usage != 'inventory':
-            res_model = 'mrp.production'
-            res_id = move.production_id.id
-            ref = move.production_id.name
-        elif move.raw_material_production_id and move.location_dest_usage != 'inventory':
-            res_model = 'mrp.production'
-            res_id = move.raw_material_production_id.id
-            ref = move.raw_material_production_id.name
-        elif move.unbuild_id:
-            res_model = 'mrp.unbuild'
-            res_id = move.unbuild_id.id
-            ref = move.unbuild_id.name
-        elif move.consume_unbuild_id:
-            res_model = 'mrp.unbuild'
-            res_id = move.consume_unbuild_id.id
-            ref = move.consume_unbuild_id.name
-        return res_model, res_id, ref
-
-    @api.model
-    def _get_linked_move_lines(self, move_line):
-        """ Return all the move lines linked to move_line.
-        parent_lines are the move lines that happen first in the production chain. We go towards the raw components.
-        children_lines are the move lines that happen later in the production chain. We go towards the final products. """
-        parent_lines, children_lines = super()._get_linked_move_lines(move_line)
-        move = move_line.move_id
-        if not parent_lines:
-            parent_lines = (move.consume_unbuild_id and move_line.produce_line_ids) or (move.production_id and move_line.consume_line_ids)
-        if not children_lines:
-            children_lines = (move.unbuild_id and move_line.consume_line_ids) or (move.raw_material_production_id and move_line.produce_line_ids)
-        return parent_lines, children_lines
